### DISCUSSION TITLE
Diff between  services.AddSingleton<ContentPart, customPart>(); and services.AddContentPart<customPart>();

### DISCUSSION THREAD

**Discussion by Ramemis** [OrchardCMS/OrchardCore#14126] [General]
I recently updated Orchard Core from version 1.0.0 to 1.4.0. While the site itself was functioning correctly after the update, I encountered an issue with the admin pages throwing a `System.InvalidCastException` on publishing contentitems. This problem was resolved when I included the custom part using `AddContentPart`. Initially, I assumed that adding `AddSingleton<ContentPart, customPart>();` would suffice, given that we were essentially resolving `customPart` as a `ContentPart`. However, the error persisted until I also added `services.AddContentPart<customPart>();`.

I'd like to understand the distinction between these two approaches. Currently, I have both the former and latter code sections in my application. My question is whether these two approaches have the same meaning and whether it's safe to remove the former (`AddSingleton`) without any potential consequences.

**Comment by MikeAlhayek** [comment]
AddContentPart manages an internal collection. Every time you register a type, we add it to an internal collection that host many parts in one collection.

**Comment by ns8482e** [comment]
I guess the services.AddContentPart is the correct way to register your parts. I believe other is for backward compatibility and deprecated
